### ISSUE TITLE
No proprietary binary files, please

### ISSUE THREAD

**Issue by msporny** [w3c/webpayments#7]
Hey @mattsaxon,

Thanks for putting the first set of flow diagrams together.

Three requests:
1. Let's not check in binary files that can easily be generated.
2. I don't know if checking in .docx files really helps us (because we can't include those in ReSpec and viewing them in the browser is problematic). My preference would be to commit SVGs or PNGs (if we absolutely have to commit something), with the preference being for SVGs (we can include SVGs in the ReSpec specs).
3. Finally, and this one is really pedantic, but can we not put spaces in directory names or file names. Many of us jump around in the command line and hack URLs to specify these directory locations and spaces frustrate some of the things we do (like having to type in %20 when specifying a directory location instead of just using a - character or a bi-capitalized character).

**Comment by mattsaxon** [comment]
1. ```
     Done
   ```
2. ```
     We don’t need the docx or even SVG/PNG at all as far as I am concerned, viewing the text (.pml) files in the tool of your choice is fine.
   ```
3. ```
     Done
   ```

From: Manu Sporny [mailto:notifications@github.com] 
Sent: 06 November 2015 20:51
To: w3c/webpayments
Cc: mattsaxon
Subject: [webpayments] No proprietary binary files, please (#7)

Hey @mattsaxon https://github.com/mattsaxon ,

Thanks for putting the repo together

Three requests:
1.  Let's not check in binary files that can easily be generated.
2.  I don't know if checking in .docx files really helps us (because we can't include those in ReSpec and viewing them in the browser is problematic). My preference would be to commit SVGs or PNGs (if we absolutely have to commit something), with the preference being for SVGs (we can include SVGs in the ReSpec specs).
3.  Finally, and this one is really pedantic, but can we not put spaces in directory names or file names. Many of us jump around in the command line and hack URLs to specify these directory locations and spaces frustrate some of the things we do (like having to type in %20 when specifying a directory location instead of just using a - character or a bi-capitalized character).

—
Reply to this email directly or view it on GitHub https://github.com/w3c/webpayments/issues/7 .  https://github.com/notifications/beacon/AK00Vwz8QoiLmcO_6yYp0SORDsSvdMluks5pDQo4gaJpZM4GdumS.gif

**Comment by msporny** [comment]
> We don’t need the docx or even SVG/PNG at all as far as I am concerned, viewing the text (.pml) files in the tool of your choice is fine.

I think it's a good idea to have the images in some graphical form (like SVG) so we can put them in a spec of some kind for those that don't want to go to the trouble of rendering the diagrams.

No need to worry about this now, the spec editors will figure the best way of integrating if/when it comes to that.

**Comment by mattsaxon** [comment]
We can probably render them direct from the text file inside an HTML page

Something like this is demoed here http://plantuml.com/jquery.html

**Comment by msporny** [comment]
> Something like this is demoed here http://plantuml.com/jquery.html

Perfect. @halindrome and I can try to figure out the details. Probably a data-include plus some post-processing hooks. I could see this becoming ReSpec feature since there are a number of groups (such as WebRTC) that have to document flows graphically in a way that is also accessible.

**Comment by msporny** [comment]
Hey @mattsaxon, I've just committed some code that will render sequence diagrams in pure Javascript and SVG.

Example of what the sequence diagram markup is here:

https://github.com/web-payments/web-payments-http-api/blob/d1d88da743f39da31d2c54a9bcc81447ebe560c5/index.html#L244-L255

The graphics generated by the markup above can be viewed here:

http://web-payments.github.io/web-payments-http-api/#issue-1

The downsides of this approach:
1. It can only be used for flow diagrams, not for complex UML class diagrams. I suggest that we try to not do the latter and assert that this basic functionality should be good enough for us.

The benefits of this approach are:
1. We don't have to check binary images into the repository for flow diagrams anymore.
2. We don't have to use .docx files anymore.
3. We could share the diagrams across different specs via ReSpec's data-include feature (I think).

**Comment by adrianhopebailie** [comment]
@msporny can you provide some guidance on how these are used?

**Comment by davidlehn** [comment]
@msporny I'd vote against the "hand drawn" mode.  It's a bit too informal.  The lines even change their wiggles every reload.  Also worth considering that it's very easy to hit the limits of simple sequence diagram tools resulting in jumping through hoops to try and describe a process.  For example loops, alt, self referencing calls, etc.  There are other js options, for example: http://jumly.tmtk.net/.  Though at some point it makes sense to look into more full UML tools.

**Comment by halindrome** [comment]
Yeah I don't love the hand drawn mode either.  I don't mind using the embedded JS to generate the drawings when we are producing drafts.  When producing static documents, however, I would prefer that we generate static SVG / PNG or whatever and reference those.  If only for Accessibility purposes.  It is very important that the illustrations be a11y friendly.

**Comment by mattsaxon** [comment]
Manu, good work on getting it to dynamically generate, though I agree the style should be the standard one rather than the hand-drawn.

What is the implementation behind this, it doesn't appear to use "jquery_plantuml.js" - is that the reason the style is different and there is a restriction to only do sequence diagrams?

**Comment by msporny** [comment]
Hey guys, after much discussion, I think we're going to fall back to the PlantUML mechanism, but use the proxy-render. 
1. We'll use the proxy render with PNG output for the wikis.
2. We'll use the proxy render with SVG output for the specs.

We may need to install our own proxy renderer, which @halindrome has already done, to make sure the images are cached and we aren't depending on the PlantUML community's proxy renderer.

**Comment by msporny** [comment]
To be clear, I'm abandoning the client-side rendering approach. The one @mattsaxon started with is good, and we've now found a good way to integrate that output into the wiki and specs.

**Comment by webpayments** [comment]
Quick follow up - plantuml.com has now been updated to support the
additional fmt parameter to the proxy.  So this now works:
http://www.plantuml.com/plantuml/proxy?fmt=svg&src=https://raw.githubusercontent.com/w3c/webpayments/gh-pages/PaymentFlows/Card/MerchantHosted-CardPayment-CGProposal.pml

On Sat, Dec 5, 2015 at 7:26 PM, Manu Sporny notifications@github.com
wrote:

> Closed #7 https://github.com/w3c/webpayments/issues/7.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/w3c/webpayments/issues/7#event-483586211.

## 

-Shane
